crates/utils/src/http_client.rs

```diff
@@ -0,0 +1,845 @@
+
+    use base64::engine::general_purpose;
+    use base64::Engine;
+    use httpmock;
+    use reqwest::header::{HeaderName, HeaderValue};
+
+    use super::*;
+
+    /// # Builder Tests
+
+    /// Verifies that HttpClient::builder creates a valid builder with the provided base URL
+    /// and all default values are properly initialized
+    #[test]
+    fn test_builder_basic_initialization() {
+        let mock_server = httpmock::MockServer::start();
+        let client = HttpClient::builder(&mock_server.base_url()).build().unwrap();
+
+        assert_eq!(client.base_url.as_str(), mock_server.base_url() + "/");
+        assert!(client.default_headers.is_empty());
+        assert!(client.default_query_params.is_empty());
+        assert!(client.default_form_data.is_empty());
+        assert!(client.default_body_params.is_empty());
+    }
+
+    /// Ensures the builder properly panics when provided with an invalid URL
+    /// Cases: malformed URLs, invalid schemes, empty URLs
+    #[test]
+    #[should_panic(expected = "Invalid base URL")]
+    fn test_builder_invalid_url() {
+        HttpClient::builder("not a url");
+    }
+
+    /// Verifies that default headers set during builder phase are:
+    /// - Properly stored in the builder
+    /// - Correctly transferred to the final client
+    /// - Applied to outgoing requests
+    #[test]
+    fn test_builder_default_headers() {
+        let mock_server = httpmock::MockServer::start();
+
+        let header_name = HeaderName::from_static("x-test");
+        let header_value = HeaderValue::from_static("test-value");
+
+        let client = HttpClient::builder(&mock_server.base_url())
+            .default_header(header_name.clone(), header_value.clone())
+            .build()
+            .unwrap();
+
+        assert!(client.default_headers.contains_key(&header_name));
+        assert_eq!(client.default_headers.get(&header_name).unwrap(), &header_value);
+    }
+
+    /// Validates default query parameter handling:
+    /// - Parameters are correctly stored
+    /// - Multiple parameters can be added
+    /// - Parameters are properly URL encoded
+    #[test]
+    fn test_builder_default_query_params() {
+        let mock_server = httpmock::MockServer::start();
+
+        let client = HttpClient::builder(&mock_server.base_url())
+            .default_query_param("key1", "value1")
+            .default_query_param("key2", "value 2") // Space should be encoded
+            .build()
+            .unwrap();
+
+        assert_eq!(client.default_query_params.len(), 2);
+        assert_eq!(client.default_query_params.get("key1").unwrap(), "value1");
+        assert_eq!(client.default_query_params.get("key2").unwrap(), "value 2");
+    }
+
+    /// Tests body parameter handling:
+    /// - Parameters are properly concatenated with '&'
+    /// - Empty parameters are handled correctly
+    /// - URL encoding is applied appropriately
+    #[test]
+    fn test_builder_default_body_params() {
```

Comment:
i don't think default body param is neede

Reply:
removed